sound/soc/sof/ipc.c

```diff
@@ -295,6 +295,9 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
 	    reply_bytes > SOF_IPC_MSG_MAX_SIZE)
 		return -ENOBUFS;
 
+	/* cancel any attempt for DSP D0i3 entry */
+	cancel_delayed_work_sync(&sdev->d0i3_work);
+
```

Comment:
handle d0i3 delayed work in generic IPC code is not very good IMO.

Reply:
@keyonjie what is not "good" about it? I am open to moving to hdev if you can give me a strong case what can go wrong if it is in sdev instead. 

Keeping it in the generic IPC TX makes it a lot more intuitive for reviewers to tell what is going on isnt? Why hide it?

Reply:
@ranj063 I mean that if I am a developer who works on IPC part only, I don't need to know how the power management(and the d0i3 delayed work) is handled, calling a helper to make sure the power is on during IPC sending is fine, but asking for taking more in my IPC code will break the decoupling discipline for different modules(e.g. for IPC and PM).

Reply:
@keyonjie IPC and PM are not different modules. They are both essential parts of the SOF core. You can't one without the other. So I cant buy the argument that someone working on IPC can be oblivious to what the DSP does in terms of PM. 

I'll argue both for and against your case here:

For: For platforms that do not support D0i3, scheduling the D0i3_work() is actually a moot point because you'd never really do anything but still keep on scheduling the workqueue for no good reason.

Against: It makes it clear as day in terms of what the expectations for opportunistic D0i3. You send the IPC and wait before entering D0i3

Now, if both you and @plbossart still feel strongly about this, I'm open to moving it.

Reply:
the bikeshed is on fire and you are arguing about the paint I am afraid.